src/gc-interface.h

```diff
@@ -207,6 +207,13 @@ JL_DLLEXPORT void *jl_gc_perm_alloc(size_t sz, int zero, unsigned align,
 //              the allocated object. All objects stored in fields of this object
 //              must be either permanently allocated or have other roots.
 struct _jl_value_t *jl_gc_permobj(size_t sz, void *ty) JL_NOTSAFEPOINT;
+// Same as jl_gc_permobj but permanently allocates a symbol (jl_sym_t). The object needs to be word aligned,
+// and tagged with jl_sym_tag. The differences between the two functions are:
+// 1. sz for this function includes the object header, and sz for jl_gc_permobj excludes the header size.
+// 2. align for this function is word align, and align for jl_gc_permobj depends on the allocation size.
+// 3. ty for this function is jl_symbol_tag << 4 and is set via jl_set_typetagof, and ty for jl_gc_permobj
+// is a datatype pointer and is set directly from the object's header.
+struct _jl_value_t *jl_gc_permsymbol(size_t sz) JL_NOTSAFEPOINT;
```

Comment:
I don't really like making this specific to symbol. If the primary difference is in the alignment, why not make a version of permobj that just takes an additional alignment arg?

Reply:
I think the alignment and size differences are actually fine, and it would be easy to integrate them with `jl_gc_permobj`. 
The problem is setting the type: because for symbols the type is set `jl_symbol_tag`, the type needs to be set atomically, whereas for objects it can be set directly. I could combine the functions but it would mean that we would have to use `jl_set_typeof` in `jl_gc_permobj`. I think it would look much cleaner but it depends on whether that's a reasonable trade-off.

Reply:
> because for symbols the type is set `jl_symbol_tag`, the type needs to be set atomically

This statement doesn't make sense to me.

Reply:
Sorry, let me try to clarify. 
The type of the symbol is set [here](https://github.com/JuliaLang/julia/blob/cf4ab8336468f73bbc768187864586bedbd59342/src/symbol.c#L42) via a call to [`jl_set_typetagof`](https://github.com/JuliaLang/julia/blob/cf4ab8336468f73bbc768187864586bedbd59342/src/julia.h#L137), whereas, for `jl_gc_permobj`, the type is set [here](https://github.com/JuliaLang/julia/blob/cf4ab8336468f73bbc768187864586bedbd59342/src/gc-stock.c#L3939) writing it directly to the header. As far as I understood from the code, that difference happens because for symbols we're setting `jl_symbol_tag`, which needs to be set atomically. So if we want to reuse `jl_gc_permobj` to allocate symbols, we need to make sure that it also sets the type atomically.

Reply:
I think it's probably just an oversight that the `header` field is not marked atomic.

Reply:
I see. I've changed the code to use `jl_gc_permobj` for symbols, and to use `jl_set_typeof` inside it. Thanks for looking into it and for the feedback :)